toolchain/base/index_base.h

```diff
@@ -52,78 +53,32 @@ struct IndexBase : public IdBase {
   using IdBase::IdBase;
 };
 
-// Equality comparison for both IdBase and IndexBase.
-template <
-    typename IndexType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr>
+// Equality comparison for both IdBase and IndexBase, including the index type
+// on either the LHS or RHS with something convertible to it. Inequality
+// automatically uses these in C++20.
+template <typename IndexType>
+  requires std::derived_from<IndexType, IdBase>
 auto operator==(IndexType lhs, IndexType rhs) -> bool {
   return lhs.index == rhs.index;
 }
-template <
-    typename IndexType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr>
-auto operator!=(IndexType lhs, IndexType rhs) -> bool {
-  return lhs.index != rhs.index;
-}
-
-template <
-    typename IndexType, typename RHSType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr,
-    typename std::enable_if_t<std::is_convertible_v<RHSType, IndexType>>* =
-        nullptr>
+template <typename IndexType, typename RHSType>
+  requires std::derived_from<IndexType, IdBase> &&
+           std::convertible_to<RHSType, IndexType>
 auto operator==(IndexType lhs, RHSType rhs) -> bool {
   return lhs.index == IndexType(rhs).index;
 }
-template <
-    typename IndexType, typename RHSType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr,
-    typename std::enable_if_t<std::is_convertible_v<RHSType, IndexType>>* =
-        nullptr>
-auto operator!=(IndexType lhs, RHSType rhs) -> bool {
-  return lhs.index != IndexType(rhs).index;
-}
-
-template <
-    typename LHSType, typename IndexType,
-    typename std::enable_if_t<std::is_base_of_v<IdBase, IndexType>>* = nullptr,
-    typename std::enable_if_t<std::is_convertible_v<LHSType, IndexType>>* =
-        nullptr>
+template <typename LHSType, typename IndexType>
+  requires std::derived_from<IndexType, IdBase> &&
+           std::convertible_to<LHSType, IndexType>
 auto operator==(LHSType lhs, IndexType rhs) -> bool {
   return IndexType(lhs).index == rhs.index;
 }
```

Comment:
```suggestion
```

We shouldn't need this; the compiler will automatically rewrite `a == b` into `b == a` if needed.

Reply:
I thought so as well, tried removing it, and got failures. 🤷

Reply:
I tried removing each separately, and it seems to still compile. Per chat with zygoloid, if you have the matching `!=` it won't (assuming you meant to use an older approach, so it won't rewrite). Maybe you checked before removing `!=`?

Reply:
Removed.